physics/discrete_trajectory_body.hpp

```diff
@@ -178,6 +183,110 @@ void DiscreteTrajectory<Frame>::ForgetBefore(Instant const& time) {
+      hint == nullptr ? nullptr : CHECK_NOTNULL(dynamic_cast<Hint*>(hint));
+  {
+    Hint* const hint = down_cast_hint;
+    // TODO(egg): do things with the hint.
+    CHECK_LE(t_min(), time);
+    CHECK_GE(t_max(), time);
+    // This is the upper bound of the interval upon which we will do the
+    // interpolation.
+    auto const upper = LowerBound(time);
+    if (upper == Begin()) {
+      return upper.degrees_of_freedom().position();
+    }
+    auto const lower = --Iterator{upper};
+    Hermite3<Instant, Position<Frame>> interpolation{
+        {lower.time(), upper.time()},
+        {lower.degrees_of_freedom().position(),
+         upper.degrees_of_freedom().position()},
+        {lower.degrees_of_freedom().velocity(),
+         upper.degrees_of_freedom().velocity()}};
+    return interpolation.Evaluate(time);
+  }
+}
+
+template<typename Frame>
+Velocity<Frame> DiscreteTrajectory<Frame>::EvaluateVelocity(
+    Instant const& time,
+    typename Trajectory<Frame>::Hint* hint) const {
+  Hint* const down_cast_hint =
+      hint == nullptr ? nullptr : CHECK_NOTNULL(dynamic_cast<Hint*>(hint));
+  {
+    Hint* const hint = down_cast_hint;
+    // TODO(egg): do things with the hint.
+    CHECK_LE(t_min(), time);
+    CHECK_GE(t_max(), time);
+    // This is the upper bound of the interval upon which we will do the
+    // interpolation.
+    auto const upper = LowerBound(time);
+    if (upper == Begin()) {
+      return upper.degrees_of_freedom().velocity();
+    }
+    auto const lower = --Iterator{upper};
+    Hermite3<Instant, Position<Frame>> interpolation{
+        {lower.time(), upper.time()},
+        {lower.degrees_of_freedom().position(),
+         upper.degrees_of_freedom().position()},
+        {lower.degrees_of_freedom().velocity(),
+         upper.degrees_of_freedom().velocity()}};
+    return interpolation.EvaluateDerivative(time);
+  }
+}
+
+template<typename Frame>
+DegreesOfFreedom<Frame> DiscreteTrajectory<Frame>::EvaluateDegreesOfFreedom(
+    Instant const& time,
+    typename Trajectory<Frame>::Hint* hint) const {
+  Hint* const down_cast_hint =
+      hint == nullptr ? nullptr : CHECK_NOTNULL(dynamic_cast<Hint*>(hint));
+  {
+    Hint* const hint = down_cast_hint;
+    // TODO(egg): do things with the hint.
+    CHECK_LE(t_min(), time);
+    CHECK_GE(t_max(), time);
+    // This is the upper bound of the interval upon which we will do the
+    // interpolation.
+    auto const upper = LowerBound(time);
+    if (upper == Begin()) {
+      return upper.degrees_of_freedom();
+    }
+    auto const lower = --Iterator{upper};
+    Hermite3<Instant, Position<Frame>> interpolation{
+        {lower.time(), upper.time()},
+        {lower.degrees_of_freedom().position(),
+         upper.degrees_of_freedom().position()},
+        {lower.degrees_of_freedom().velocity(),
+         upper.degrees_of_freedom().velocity()}};
+    return {interpolation.Evaluate(time),
+            interpolation.EvaluateDerivative(time)};
```

Comment:
Could we please factor out this triplication by having a function give-me-the-Hermite-polynomial for the time and the hint?